library/pkcs5.c

```diff
@@ -152,6 +152,7 @@ int mbedtls_pkcs5_pbes2_ext(const mbedtls_asn1_buf *pbe_params, int mode,
     mbedtls_cipher_type_t cipher_alg;
     mbedtls_cipher_context_t cipher_ctx;
     unsigned int padlen = 0;
+    mbedtls_cipher_padding_t padding;
```

Comment:
So I don't like having these at the top of the function, since it's only used in a very small locus within this function. It feels like a block at that point would be neater, but that's a lot of churn for a compiler fix. Or could that code be moved into a separate function?

If you think this change is best, however, I won't oppose it

Reply:
(obviously same comment for `library/pkcs12.c`)

Reply:
I think this is not worth the churn - IMO neither option is obviously better. Although using a block does save 12b in x509_create.c ...

Reply:
Separating out into a function does reduce the cyclomatic complexity of `mbedtls_pkcs5_pbes2_ext()`

```
diff --git a/library/pkcs5.c b/library/pkcs5.c
index 2756d058e..c2f583ae3 100644
--- a/library/pkcs5.c
+++ b/library/pkcs5.c
@@ -136,6 +136,33 @@ int mbedtls_pkcs5_pbes2(const mbedtls_asn1_buf *pbe_params, int mode,
 }
 #endif
 
+#if defined(MBEDTLS_CIPHER_MODE_WITH_PADDING)
+static int mbedtls_pkcs5_pbes2_set_padding(mbedtls_cipher_context_t *cipher_ctx,
+                                           int mode)
+{
+    /* PKCS5 uses CBC with PKCS7 padding (which is the same as
+     * "PKCS5 padding" except that it's typically only called PKCS5
+     * with 64-bit-block ciphers).
+     */
+    mbedtls_cipher_padding_t padding = MBEDTLS_PADDING_PKCS7;
+
+#if !defined(MBEDTLS_CIPHER_PADDING_PKCS7)
+    /* For historical reasons, when decrypting, this function works when
+     * decrypting even when support for PKCS7 padding is disabled. In this
+     * case, it ignores the padding, and so will never report a
+     * password mismatch.
+     */
+    if (mode == MBEDTLS_DECRYPT) {
+        padding = MBEDTLS_PADDING_NONE;
+    }
+#else
+    (void) mode;
+#endif
+
+    return mbedtls_cipher_set_padding_mode(cipher_ctx, padding);
+}
+#endif /* MBEDTLS_CIPHER_MODE_WITH_PADDING */
+
 int mbedtls_pkcs5_pbes2_ext(const mbedtls_asn1_buf *pbe_params, int mode,
                             const unsigned char *pwd,  size_t pwdlen,
                             const unsigned char *data, size_t datalen,
@@ -242,25 +269,11 @@ int mbedtls_pkcs5_pbes2_ext(const mbedtls_asn1_buf *pbe_params, int mode,
     }
 
 #if defined(MBEDTLS_CIPHER_MODE_WITH_PADDING)
-    /* PKCS5 uses CBC with PKCS7 padding (which is the same as
-     * "PKCS5 padding" except that it's typically only called PKCS5
-     * with 64-bit-block ciphers).
-     */
-    mbedtls_cipher_padding_t padding = MBEDTLS_PADDING_PKCS7;
-#if !defined(MBEDTLS_CIPHER_PADDING_PKCS7)
-    /* For historical reasons, when decrypting, this function works when
-     * decrypting even when support for PKCS7 padding is disabled. In this
-     * case, it ignores the padding, and so will never report a
-     * password mismatch.
-     */
-    if (mode == MBEDTLS_DECRYPT) {
-        padding = MBEDTLS_PADDING_NONE;
-    }
-#endif
-    if ((ret = mbedtls_cipher_set_padding_mode(&cipher_ctx, padding)) != 0) {
+    if ((ret = mbedtls_pkcs5_pbes2_set_padding(cipher_ctx, mode)) != 0) {
         goto exit;
     }
 #endif /* MBEDTLS_CIPHER_MODE_WITH_PADDING */
+
     if ((ret = mbedtls_cipher_crypt(&cipher_ctx, iv, enc_scheme_params.len,
                                     data, datalen, output, output_len)) != 0) {
         ret = MBEDTLS_ERR_PKCS5_PASSWORD_MISMATCH;
```

Reply:
but it's a lot of churn